drivers/partition/partition.c

```diff
@@ -0,0 +1,255 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <debug.h>
+#include <io_storage.h>
+#include <partition.h>
+#include <platform.h>
+#include <string.h>
+
+static part_entry_t entry_list[MAX_PARTITION_NUM];
+static int part_num_entries;
+static partition_ops_t part_ops;
+static uint8_t mbr_sector[PARTITION_BLOCK_SIZE];
+
+static int unicode_to_ascii(uint16_t *str_in, uint8_t *str_out)
+{
+	uint8_t *name = (uint8_t *)str_in;
+	int i;
+
+	assert((name[0] != '\0') && (str_in != NULL) && (str_out != NULL));
+
+	/* Calculate the unicode string length. */
+	for (i = 1; i < (EFI_NAMELEN << 1); i += 2) {
+		if (name[i] != '\0')
+			return -EINVAL;
+	}
+	/* Convert */
+	for (i = 0; i < (EFI_NAMELEN << 1); i += 2) {
+		str_out[i >> 1] = name[i];
+		if (name[i] == '\0')
+			break;
+	}
+	return 0;
+}
+
+static int parse_gpt_entry(gpt_entry_t *gpt_entry, int entry_index)
+{
+	int result;
+
+	assert((gpt_entry != 0) && (entry_index >= 0) &&
+	       (entry_index < MAX_PARTITION_NUM));
+
+	/* exhaused partition entry */
+	if ((gpt_entry->first_lba == 0) && (gpt_entry->last_lba == 0)) {
+		return -EINVAL;
+	}
+
+	memset(&entry_list[entry_index], 0, sizeof(part_entry_t));
+	result = unicode_to_ascii(gpt_entry->name,
+				  (uint8_t *)entry_list[entry_index].name);
+	if (result != 0) {
+		return result;
+	}
+	entry_list[entry_index].start = (uint64_t)gpt_entry->first_lba *
+					PARTITION_BLOCK_SIZE;
+	entry_list[entry_index].length = (uint64_t)(gpt_entry->last_lba -
+						    gpt_entry->first_lba + 1) *
+					 PARTITION_BLOCK_SIZE;
+	return 0;
+}
+
+static int verify_partition_mbr(uintptr_t image_handle)
+{
+	size_t bytes_read;
+	mbr_entry_t *entry;
+	int result;
+
+	/* MBR partition table is in LBA0. */
+	result = io_seek(image_handle, IO_SEEK_SET, MBR_OFFSET);
+	if (result != 0) {
+		WARN("Failed to seek (%i)\n", result);
+		return result;
+	}
+	result = io_read(image_handle, (uintptr_t)&mbr_sector,
+			 PARTITION_BLOCK_SIZE, &bytes_read);
+	if (result != 0) {
+		WARN("Failed to read data (%i)\n", result);
+		return result;
+	}
+
+	/* Check MBR boot signature. */
+	if ((mbr_sector[PARTITION_BLOCK_SIZE - 2] != 0x55) ||
+	    (mbr_sector[PARTITION_BLOCK_SIZE - 1] != 0xAA)) {
+		return -EINVAL;
+	}
+	/* Check the first partition entry. */
+	entry = (mbr_entry_t *)((uintptr_t)&mbr_sector +
+				MBR_PRIMARY_ENTRY_OFFSET);
+	if (entry->type == PARTITION_TYPE_GPT)
+		return -EAGAIN;
+	return -EINVAL;
+}
+
+static void dump_entries(int num)
```

Comment:
This whole function body should be wrapped in a `#if LOG_LEVEL >= LOG_LEVEL_INFO` block.